src/core/toPath.js

```diff
@@ -102,139 +102,70 @@ const allowingArrays = fn => arg => {
-        if (str.charAt(index) === ']') index++
-
-        // Stop if end of str has been reached
-        if (index === str.length) break
-
-        // If next character is a dot, move index after it (skip it)
-        if (str.charAt(index) === '.') index++
-
-      } else { // If array index is not a quoted string
-
-        // Look for the closing square bracket
-        const closingBracketIndex = str.indexOf(']', index)
-
-        // If no closing bracket found, stop if end of str is reached, or continue to next iteration
-        if (closingBracketIndex === -1) {
-          if (index !== str.length) path.push(str.substring(index))
-          break
-        }
-
-        // Fetch the content of brackets and move index after closing bracket
-        const arrayIndexValue = str.substring(index, closingBracketIndex)
-        index = closingBracketIndex + 1
-
-        // If next character is a dot, move index after it (skip it)
-        if (str.charAt(index) === '.') index++
-
-        // Shorthand: if array index is the whole slice add it to path
-        if (arrayIndexValue === ':') {
-          path.push([undefined, undefined])
-        } else {
-
-          // Look for a slice quote
-          const sliceDelimIndex = arrayIndexValue.indexOf(':')
-
-          // If no slice quote found
-          if (sliceDelimIndex === -1) {
-            // Parse array index as a number
-            const nArrayIndexValue = Number(arrayIndexValue)
-
-            // Add array index to path, either as a valid index (positive int), or as a string
-            path.push(isIndex(nArrayIndexValue) ? nArrayIndexValue : arrayIndexValue)
+  const path = stringToPath2(str)
+  return str[0] === '.' ? ['', ...path] : path
+}
 
-          } else { // If a slice quote is found
+const splitAtFirstOccurence = (str, separators) => {
+  const partitionIndex = separators
+    .map(separator => str.indexOf(separator))
+    .map(index => index >= 0 ? index : str.length)
+    .reduce((minIndex, index) => Math.min(minIndex, index))
+  return [str.substring(0, partitionIndex), str.substr(partitionIndex, 1), str.substring(partitionIndex + 1)]
+}
 
-            // Fetch slice start and end, and parse them as slice indexes (empty or valid int)
-            const sliceStart = arrayIndexValue.substring(0, sliceDelimIndex), sliceEnd = arrayIndexValue.substring(sliceDelimIndex + 1)
-            const nSliceStart = toSliceIndex(sliceStart), nSliceEnd = toSliceIndex(sliceEnd)
+/**
+ * 
+ * @param {string} str string to parse, expected to start with an opening square bracket followed by a quote char
+ * @param {*} quote the quote char
+ * @returns {[string, string]} a tuple of the dequoted path segment and the rest of the input string
+ * @example parseQuotedBracketNotation('["abc"].def', '"') // ['abc', 'def']
+ * @example parseQuotedBracketNotation('["abc', '"') // ['abc', '']
+ * @example parseQuotedBracketNotation('abc', '"') // ['c', '']
```

Comment:
Is this a typo ? Why would 'c' end up in the dequoted segment path ?

Reply:
OK this is because of the substring(2), but is this correct ?

Reply:
Correct.